### PR TITLE
feat(core): v3 core solution

### PR DESCRIPTION
**Pull request by jayzhang**
1. improve V3 APIs 
 - packages/api/src/v3/plugins.ts
 - packages/api/src/v3/solution.ts
2. implement FXCore V3 APIs, upgrade questionModelMW to support V3
 - packages/fx-core/src/core/index.ts
 - packages/fx-core/src/core/middleware/questionModel.ts
3. implement part of azure solution v3 APIs
 - packages/fx-core/src/plugins/solution/fx-solution/v3/addResource.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/constants.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/init.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/provision.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/provisionLocal.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/questions.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/scaffold.ts
 - packages/fx-core/src/plugins/solution/fx-solution/v3/solution.ts
4. adjust SPFx solution name
 - packages/fx-core/src/plugins/solution/spfx-solution/scaffolding.ts
 - packages/fx-core/src/plugins/solution/spfx-solution/solution.ts
 - packages/fx-core/src/plugins/solution/spfx-solution/init.ts
5. upgrade arm module to support V3 APIs
 - packages/fx-core/src/plugins/solution/fx-solution/arm.ts
 - packages/fx-core/src/plugins/solution/fx-solution/utils/depsChecker/bicepChecker.ts
6. For v3 implement, VS and VS Code have the same behaviors for create project questions (different from CLI and CLI_HELP)

### CODE REVIEW

**Review by LongOddCode — APPROVED** [review]

File: packages/cli/.vscode/launch.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
       "name": "Launch new command (non-interactive)",
       "skipFiles": ["<node_internals>/**"],
       "program": "${workspaceFolder}/cli.js",
-      "args": ["new", "--interactive", "false"],
+      "args": ["new", "--interactive", "false", "--app-name", "fxE2E1639996799061b9", "--capabilities", "tab", "bot", "--azure-resources", "function", "--programming-language", "typescript"] ,
```
Comment (wenytang-ms) [inline packages/cli/.vscode/launch.json:13]: please verify this value is need to check in to source branch

Comment (jayzhang) [reply packages/cli/.vscode/launch.json]: reverted

**Review by LongOddCode — APPROVED** [review]

**Comment by jayzhang** [comment]
E2E test is pass: https://github.com/OfficeDev/TeamsFx/actions/runs/1604951961

File: packages/fx-core/src/plugins/solution/fx-solution/utils/depsChecker/bicepChecker.ts [hunk 2]
```diff
@@ -360,20 +367,17 @@ function getCommonProps(): { [key: string]: string } {
...
-  ctx: SolutionContext
+  ui?: UserInteraction,
+  telemetryReporter?: TelemetryReporter
 ): Promise<boolean> {
-  if (!ctx.ui) {
-    // no dialog, always continue
-    return true;
-  }
-  const res = await ctx.ui?.showMessage("info", message, true, Messages.learnMoreButtonText);
+  const res = await ui?.showMessage("info", message, true, Messages.learnMoreButtonText);
```
Comment (a1exwang) [inline packages/fx-core/src/plugins/solution/fx-solution/utils/depsChecker/bicepChecker.ts:373]: It seems your refactor changes the original logic. When `ui === undefined`, the new logic will return `false` (and send telemetry) but the old logic will return `true`.

Comment (jayzhang) [reply packages/fx-core/src/plugins/solution/fx-solution/utils/depsChecker/bicepChecker.ts:377]: fixed

**Review by a1exwang — APPROVED** [review]
approve for the env checker part

**Review by meifans — APPROVED** [review]

**Review by LongOddCode — APPROVED** [review]

**Review by a1exwang — APPROVED** [review]

**Pull request merged**